### PR TITLE
[cpp_on_shore_task_3] rb-tree learning

### PR DESCRIPTION
**Pull request by SheldonLoveCoding**
### What?

- brief: 阅读STL tree的源码(stl_tree.h 和 tree.cc)，学习红黑树的原理和使用
- author: liuxd_sheldon@163.com

### Directory Specification

```
├─cpp_on_shore
    │  ├─chris
    │  │  ├─doc
    │  │  │  └─task_3
    │  │  └─src
    │  │      └─task_3
    │  ├─roger_lv
    │  │  ├─doc
    │  │  │  └─task_3
    │  │  └─src
    │  │      └─task_3
    │  └─sheldon
    │      ├─doc
    │      │  └─task_3
    │      └─src
    │          └─task_3
```

### What changes are included in this PR?

按照要求创建task_3学习目录，分别创建doc和src文件夹，阅读stl_tree.h。保存task_3各自笔记和代码

### Are these changes tested?

编写的代码经过测试

### CODE REVIEW

File: stl/cpp_on_shore/chris/src/task_3/rb_tree.h [hunk 1]
```diff
@@ -0,0 +1,353 @@
+#include <assert.h>
```
Comment (Light-City) [inline stl/cpp_on_shore/chris/src/task_3/rb_tree.h:1]: cassert

File: stl/cpp_on_shore/chris/src/task_3/rb_tree.h [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+  RbTreeNode* right;   // 右子节点
+  RbTreeNode* parent;  // 父节点
+};
+
+template <typename Type>
+class RbTree {
+ public:
+  // constructor
+  RbTree() {
+    Nil = BuyNode();
```
Comment (Light-City) [inline stl/cpp_on_shore/chris/src/task_3/rb_tree.h:22]: BuyNode?

File: stl/cpp_on_shore/chris/src/task_3/rb_tree.h [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+  Color color;         // 颜色
+  Type key;            // 关键字
+  RbTreeNode* left;    // 左子节点
+  RbTreeNode* right;   // 右子节点
+  RbTreeNode* parent;  // 父节点
+};
+
+template <typename Type>
+class RbTree {
+ public:
```
Comment (Light-City) [inline stl/cpp_on_shore/chris/src/task_3/rb_tree.h:19]: split node class and rbtree class may be a good design

File: stl/cpp_on_shore/chris/src/task_3/rb_tree.h [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+    std::cout << node->key << "(" << ((node->color == BLACK) ? "BLACK" : "RED") << ")"
+              << " ";
+    if (node->right != Nil) {
+      InOrderPrint(node->right);
+    }
+  }
+
+ private:
+  RbTreeNode<Type>* root;  // 根指针
+  RbTreeNode<Type>* Nil;   // 外部节点，表示空节点，颜色为黑色
```
Comment (Light-City) [inline stl/cpp_on_shore/chris/src/task_3/rb_tree.h:352]: nil

File: stl/cpp_on_shore/sheldon/src/task_3/my_set.hpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+ */
+#include <bits/concept_check.h>
+#include <bits/stl_tree.h>
+#include <initializer_list>
+
+template <typename _Key, typename _Compare = std::less<_Key>,
+          typename _Alloc = std::allocator<_Key>>
+class my_set {
+ private:
+  std::_Rb_tree<_Key, _Key, std::_Identity<_Key>, _Compare, _Alloc> itree;
```
Comment (Light-City) [inline stl/cpp_on_shore/sheldon/src/task_3/my_set.hpp:12]: rbtree

File: stl/cpp_on_shore/sheldon/src/task_3/my_set.hpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
+/*
+ * 利用stl_tree.h实现set的自定义接口
+ */
+#include <bits/concept_check.h>
```
Comment (Light-City) [inline stl/cpp_on_shore/sheldon/src/task_3/my_set.hpp:4]: where using?

File: stl/cpp_on_shore/chris/src/task_3/rb_tree.h (hunk 1 above)
Comment (ChristopherCao) [reply stl/cpp_on_shore/chris/src/task_3/rb_tree.h:1]: Copy that, already corrected.

File: stl/cpp_on_shore/chris/src/task_3/rb_tree.h (hunk 2 above)
Comment (ChristopherCao) [reply stl/cpp_on_shore/chris/src/task_3/rb_tree.h:22]: Now I use a more objective name for this func, called ApplyNode.

File: stl/cpp_on_shore/chris/src/task_3/rb_tree.h (hunk 3 above)
Comment (ChristopherCao) [reply stl/cpp_on_shore/chris/src/task_3/rb_tree.h:19]: yes, I think I have already splited the two classes in my original design. One is the RbTreeNode class, the other is the RbTree.

File: stl/cpp_on_shore/chris/src/task_3/rb_tree.h (hunk 4 above)
Comment (ChristopherCao) [reply stl/cpp_on_shore/chris/src/task_3/rb_tree.h:352]: Copy that. Already corrected.

**Review by Light-City — COMMENTED** [review]
LGTM

**Pull request merged**